blockstore/badger/versions/badger_v4.go

```diff
@@ -0,0 +1,277 @@
+}
+
+func (wb *BadgerV4WriteBatch) Delete(key []byte) error {
+	return wb.WriteBatch.Delete(key)
+}
+
+func (wb *BadgerV4WriteBatch) Flush() error {
+	return wb.WriteBatch.Flush()
+}
+
+func (wb *BadgerV4WriteBatch) Cancel() {
+	wb.WriteBatch.Cancel()
+}
+
+type BadgerV4Stream struct {
+	*badger.Stream
+}
+
+func (s *BadgerV4Stream) SetNumGo(numGo int) {
+	s.Stream.NumGo = numGo
+}
+
+func (s *BadgerV4Stream) SetLogPrefix(prefix string) {
+	s.Stream.LogPrefix = prefix
+}
+func (s *BadgerV4Stream) ForEach(ctx context.Context, fn func(key string, value string) error) error {
+	s.Stream.Send = func(buf *z.Buffer) error {
+		list, err := badger.BufferToKVList(buf)
+		if err != nil {
+			return fmt.Errorf("buffer to KV list conversion: %w", err)
+		}
+		for _, kv := range list.Kv {
+			if kv.Key == nil || kv.Value == nil {
+				continue
+			}
+			err := fn(string(kv.Key), string(kv.Value))
+			if err != nil {
+				return xerrors.Errorf("foreach function: %w", err)
+			}
+
+		}
+		return nil
+	}
+	if err := s.Orchestrate(ctx); err != nil {
+		return xerrors.Errorf("orchestrate stream: %w", err)
+	}
+	return nil
+}
+
+func (s *BadgerV4Stream) Orchestrate(ctx context.Context) error {
+	return s.Stream.Orchestrate(ctx)
+}
+
+type BadgerV4Txn struct {
+	*badger.Txn
+}
+
+func (txn *BadgerV4Txn) Get(key []byte) (Item, error) {
+	item, err := txn.Txn.Get(key)
+	return &BadgerV4Item{item}, err
+}
+
+func (txn *BadgerV4Txn) Set(key, val []byte) error {
+	return txn.Txn.Set(key, val)
+}
+
+func (txn *BadgerV4Txn) Delete(key []byte) error {
+	return txn.Txn.Delete(key)
+}
+
+func (txn *BadgerV4Txn) Commit() error {
+	return txn.Txn.Commit()
+}
+
+func (txn *BadgerV4Txn) Discard() {
+	txn.Txn.Discard()
+}
+
+func (txn *BadgerV4Txn) NewIterator(opts IteratorOptions) Iterator {
+	badgerOpts := badger.DefaultIteratorOptions
+	badgerOpts.PrefetchSize = opts.PrefetchSize
+	badgerOpts.Prefix = opts.Prefix
+	return &BadgerV4Iterator{txn.Txn.NewIterator(badgerOpts)}
+}
+
+type BadgerV4Iterator struct {
+	*badger.Iterator
+}
+
+func (it *BadgerV4Iterator) Next()           { it.Iterator.Next() }
+func (it *BadgerV4Iterator) Rewind()         { it.Iterator.Rewind() }
+func (it *BadgerV4Iterator) Seek(key []byte) { it.Iterator.Seek(key) }
+func (it *BadgerV4Iterator) Close()          { it.Iterator.Close() }
+func (it *BadgerV4Iterator) Item() Item      { return &BadgerV4Item{it.Iterator.Item()} }
+func (it *BadgerV4Iterator) Valid() bool     { return it.Iterator.Valid() }
+
+type BadgerV4Item struct {
+	*badger.Item
+}
+
+func (item *BadgerV4Item) Value(fn func([]byte) error) error {
+	return item.Item.Value(fn)
+}
+
+func (item *BadgerV4Item) Key() []byte {
+	return item.Item.Key()
+}
+
+func (item *BadgerV4Item) Version() uint64 {
+	return item.Item.Version()
+}
+
+func (item *BadgerV4Item) ValueCopy(dst []byte) ([]byte, error) {
+	return item.Item.ValueCopy(dst)
+}
+
+func (item *BadgerV4Item) ValueSize() int64 {
+	return item.Item.ValueSize()
+}
```

Comment:
We're already embedding `Item`. Do we need to re-define these functions as well? I think we already meet the required interface, right?

Reply:
(this and the other cases)

Reply:
I noticed that [Item.Version()](https://pkg.go.dev/github.com/dgraph-io/badger/v4#Item.Version) was unused and I removed it. 6d5d67918. 

Item is used in blockstore when we use the badger iterator. Item.Key for example is used in `Blockstore.AllKeysChan ()` as

```
for iter.Rewind(); iter.Valid(); iter.Next() {
    ...
    k := iter.Item().Key()
    ...
}
```
 I think we need a specific function for v2 Item and v4 Item as we have it in this PR.